Add hidden and sub_command_help options

WilliamMcCumstie opened the pull request:
Their are two major changes in this PR that will allow sub command specific help to be created. Please refer to commits for major discussion of changes.

The first change is the introduction of `hidden` commands. When a `Command` is created, if the the hidden method is set to a truthie value, then it will not appear in the general help for the whole command. It will however still have specific command help and may have any action assigned to it.

The second change is `sub_command_help`. If `sub_command_help` is set to a truthie value then instead of rendering the command specific help, a subcommand help page is displayed. It is smilar to the general help, however only lists the subcommands. By default, an action is included that will trigger the help to be displayed even if `--help` is not include. Unfortunately due to the coupling betweeen `Runner` and `Command` objects, the only way I could do this is by exec the command again with the `--help` flag.

WilliamMcCumstie committed:
Add option to hide commands from general help

Adds a hidden method to Commands which prevents the command from
appearing in the general help. The hidden command will still have
a command specific help. The intended use is to hide the base component
of sub commands. However still have the base component respond to -h

WilliamMcCumstie committed:
Create subcommand action that renders sub help

A new Commander::Command method `sub_command_help` has been included
that causes the subcommand help to be rendered. Now when a subcommand
is called, instead of rendering command_help.erb, subcommand_help.erb
is rendered instead. The sub_command_method by default will rexec with
the help flag as its action. Note this action can be overridden.

Runner had to be modified to detect if a Command has the sub_command_help
flag. If so, only subcommands of Command are passed into the templater.
All other commands are removed from the scope.

Help_Formatter had to be modified to allow the Command binding
to be merged with the Runner binding. This is because the full list of
subcommands are in the Runner scope where the active commands info is
in the Command scope.

Helper_Formatter::Terminal has been modified to specificly merge the Command
and Runner bindings and render the subcommand help.

Two new erb templates had to be created for the subcommand help.

bobwhitelock commented:
@WilliamMcCumstie - this looks great to me, thanks!

Only suggestion I would make is maybe changing the options to be called something like `sub_command` for `hidden` and `sub_command_group` for `sub_command_help`. This might better document the purpose of adding these options when specifying a command, as it indicates why you'd add them rather than what they do, and means they'd be appropriately named if we later want to add more special behaviour for sub commands or sub command groups. What do you think?

WilliamMcCumstie committed:
Rename sub command methods for clarity

The `hidden` method has been aliased as `sub_command`. This way a
command can be marked as a subcommand in the cli making it clearer
what it's functionality is. It will also allow additional features
to be added to sub_commands but not hidden into the future.

Sub_command_help has been renamed to sub_command_group as well to mark
that it applies to the group command and not the actual sub_command.
